[JIT] API for finding a common ancestor block for a pair of nodes

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29261 [JIT] Trace module calls
* **#28864 [JIT] API for finding a common ancestor block for a pair of nodes**

Differential Revision: [D18219786](https://our.internmc.facebook.com/intern/diff/D18219786)

Commit:
[JIT] API for finding a common ancestor block for a pair of nodes

[ghstack-poisoned]

resistor commented on torch/csrc/jit/ir.cpp:
```diff
@@ -1328,6 +1328,23 @@ void Node::removeFromList() {
 }
 
+Block* Node::findCommonAncestorBlockWith(Node* n) {
+  std::unordered_set<const Block*> seen_blocks_rhs;
```
It'd be nice to have a fast exit when both nodes are in the same block.

resistor commented on torch/csrc/jit/ir.cpp:
```diff
@@ -1328,6 +1328,23 @@ void Node::removeFromList() {
+Block* Node::findCommonAncestorBlockWith(Node* n) {
+  std::unordered_set<const Block*> seen_blocks_rhs;
+
+  for (Block* rhs_block = n->owningBlock(); rhs_block;
```
I believe a co-iteration over both chains simultaneously would be more efficient in most cases, because you'll end up not having to construct the full set.

eellison replied:
I believe this function does it correctly: https://github.com/pytorch/pytorch/blob/f7b2778cb1db148f1c901ea08abcbfa0e0a78f44/torch/csrc/jit/passes/canonicalize.cpp#L91

Commit:
Update on "[JIT] API for finding a common ancestor block for a pair of nodes"


Differential Revision: [D18219786](https://our.internmc.facebook.com/intern/diff/D18219786)

[ghstack-poisoned]

suo reviewed (dismissed):
Can we add some tests for this functionality? The IR Parser should make it not terrible

suo commented on torch/csrc/jit/ir.h:
```diff
@@ -736,7 +736,9 @@ struct TORCH_API Node {
   }
 
-private:
+  Block* findCommonAncestorBlockWith(Node* n);
```
There are also a number of places where we do stuff like this, it could be nice to clean them up by using this API. Topological moves come to mind

suo replied:
And…DCE? maybe? I don't remember

eellison replied:
This function could be refactored to using this API: https://github.com/pytorch/pytorch/blob/4bfe2f0900a602017a31c0aaa8f00750412a2d83/torch/csrc/jit/ir.cpp#L1176

jamesr66a replied:
Hmm, I tried using this API for that `isBeforeOrAfter` but `isBeforeOrAfter` requires having pointers to the ancestors Nodes that reside in the common ancestor block

eellison replied:
Sorry i meant to say `isBeforeOrAfter` could be refactored using this function. but actually i take it back

Commit:
Update on "[JIT] API for finding a common ancestor block for a pair of nodes"


Differential Revision: [D18219786](https://our.internmc.facebook.com/intern/diff/D18219786)

[ghstack-poisoned]

Commit:
Update on "[JIT] API for finding a common ancestor block for a pair of nodes"


Differential Revision: [D18219786](https://our.internmc.facebook.com/intern/diff/D18219786)

[ghstack-poisoned]

Commit:
Update on "[JIT] API for finding a common ancestor block for a pair of nodes"


Differential Revision: [D18219786](https://our.internmc.facebook.com/intern/diff/D18219786)

[ghstack-poisoned]

eellison commented on test/cpp/jit/test_ir.cpp:
```diff
@@ -92,5 +92,88 @@ void testBlocks() {
+  torch::jit::Graph g;
+  torch::jit::script::parseIR(input_str, &g);
+
+  std::vector<std::string> value_names{"6", "7", "9", "10"};
```
`parseIR` has another API that fills out a map with string, IValue * pairs:
`parseIR(
    const std::string& str,
    torch::jit::Graph* graph,
    std::unordered_map<std::string, Value*>& vmap)` , can you use that instead of `populate_values`

eellison commented on test/cpp/jit/test_ir.cpp:
```diff
@@ -92,5 +92,88 @@ void testBlocks() {
+  ASSERT_EQ(name_to_value.size(), 4);
+
+  /* clang-format off */
+  int ref_blocks_from_graph[4][4] = {
```
Encoding expected results like this works, another alternative is to include your previous algorithm in this test file and check against that.

eellison commented on test/cpp/jit/test_ir.cpp:
```diff
@@ -92,5 +92,88 @@ void testBlocks() {
+      Value* j_val = name_to_value[value_names[j]];
+      Block* common_ancestor =
+          i_val->node()->findCommonAncestorBlockWith(j_val->node());
+      int blocks_from_graph_block = 0;
```
Nit: can just do 
`blocks_from_graph_block =
            common_ancestor->param_node()->blocksFromGraphBlock();`
don't need the if condition

eellison commented on torch/csrc/jit/ir.h:
```diff
@@ -932,6 +936,13 @@ struct Block {
   TORCH_API void cloneFrom(Block* src, std::function<Value*(Value*)> value_map);
   TORCH_API void remapTypes(const std::function<TypePtr(TypePtr)>& type_map);
 
+  Block* parentBlock() {
```
This isn't called anywhere

eellison reviewed (approved):
Looks good, a few comments / nits

Commit:
Update on "[JIT] API for finding a common ancestor block for a pair of nodes"


Differential Revision: [D18219786](https://our.internmc.facebook.com/intern/diff/D18219786)

[ghstack-poisoned]
